### PR TITLE
support keras in distributed tensorflow on analytics-zoo

### PR DESCRIPTION
**Pull request by yangw1234**
API:
```python

   # 1. construct tf dataset
   dataset = ...

   # 2. construct keras model
    data = Input(shape=[784])
    x = Dense(64, activation='relu')(x)
    predictions = Dense(10, activation='softmax')(x)

    model = Model(input=data, output=predictions)
    model.compile(optimizer='rmsprop',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
   # 3. create TFOptimizer and train
    opt = TFOptimizer.from_keras_model(model, dataset)
    opt.optimize(end_trigger=MaxEpoch(max_epoch))
   
```

### CODE REVIEW

**Comment by jason-dai** [comment]
1) Don't change the existing interface
2) TFOptimizer.from_keras(model, ...)
3) What if the arguments model.comile are not supported in BigDL?

**Comment by yangw1234** [comment]
@jason-dai 
1. changed back the interface
2. done
3. if some arguments are not supported, an error will be raised to indicate that we do not support that.

File: pyzoo/zoo/pipeline/api/net.py [hunk 1]
```diff
@@ -30,7 +32,8 @@
     JavaValue, get_node_and_core_number
 from zoo.common import Sample, JTensor
 from zoo.feature.image import ImageSet
-from zoo.pipeline.api.keras.engine.topology import ZooKerasLayer, KerasNet
+from zoo.pipeline.api.keras.engine.topology import ZooKerasLayer, KerasNet, SGD, RMSprop, Adagrad, Adadelta, Adam, \
+    Adamax, Top1Accuracy, to_bigdl_metric, Loss
```
Comment (hkvision) [inline pyzoo/zoo/pipeline/api/net.py]: Just change to import *? strange to see import SGD from topology...

File: pyzoo/zoo/pipeline/api/net.py [hunk 2]
```diff
@@ -441,13 +444,131 @@ def __init__(self, loss, optim_method, sess=None,
...
+
+        return optim_method
+
+    @staticmethod
+    def _to_bigdl_optim_method_from_tf_optimizer(tf_optimizer):
+        raise NotImplementedError()
+
+
+    @staticmethod
+    def _to_bigdl_optim_method_from_keras_optimizer(keras_optimizer):
```
Comment (hkvision) [inline pyzoo/zoo/pipeline/api/net.py]: is this the same as https://github.com/intel-analytics/BigDL/blob/master/pyspark/bigdl/keras/optimization.py#L77?
Or you want to write it here to add flexibility for possible bug in BigDL?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala [hunk 3]
```diff
@@ -188,7 +194,7 @@ class TFValidationMethod(val valMethod: ValidationMethod[Float],
       outputs
     }
 
-    val to1basedLabel =
+    val to1basedLabel = !valMethod.isInstanceOf[Accuracy[Float]]
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: &&

File: pyzoo/zoo/pipeline/api/net.py [hunk 4]
```diff
@@ -21,6 +21,8 @@
 import os
 import json
 import numpy as np
+import keras
+import keras.backend as keras_backend
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: 1) Don't import these unconditionally
2) import tensorflow.keras?

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+slim = tf.contrib.slim
+
+
+def main(max_epoch, data_num):
+    sc = init_nncontext()
+
+    # get data, pre-process and create TFDataset
+    def get_data_rdd(dataset):
+        (images_data, labels_data) = mnist.read_data_sets("/tmp/mnist", dataset)
```
[the remark this answers is not in the document]
Comment (yangw1234) [reply pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: this reads as 0-based label

File: pyzoo/zoo/pipeline/api/net.py (hunk 2 above)
Comment (yangw1234) [reply pyzoo/zoo/pipeline/api/net.py]: did not know there is one already, I'll change to bigdl's

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hkvision) [reply pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: okay

File: pyzoo/zoo/pipeline/api/net.py (hunk 2 above)
Comment (hkvision) [reply pyzoo/zoo/pipeline/api/net.py]: Sure. If you find the mapping in BigDL has possible bugs or you want to modify, then we can add here :).

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+from bigdl.optim.optimizer import *
+import numpy as np
+import sys
+from keras.models import Model
+from keras.layers import *
+
+from bigdl.dataset import mnist
+from bigdl.dataset.transformer import *
+
+sys.path.append("/tmp/models/slim")  # add the slim library
```
Comment (hkvision) [inline pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: Any instructions in README to guide users to put the slim path under tmp?

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    x = Dense(64, activation='relu')(x)
+    predictions = Dense(10, activation='softmax')(x)
+
+    model = Model(input=data, output=predictions)
+
+    model.compile(optimizer='rmsprop',
+                  loss='sparse_categorical_crossentropy',
+                  metrics=['accuracy'])
+
+    # (images_data, labels_data) = mnist.read_data_sets("/tmp/mnist", "train")
```
Comment (hkvision) [inline pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: remove these lines if they are not needed.

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+from keras.models import Model
+from keras.layers import *
+
+from bigdl.dataset import mnist
+from bigdl.dataset.transformer import *
+
+sys.path.append("/tmp/models/slim")  # add the slim library
+from nets import lenet
+
+slim = tf.contrib.slim
```
Comment (hkvision) [inline pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: actually is slim used in this example?

File: pyzoo/zoo/pipeline/api/net.py (hunk 1 above)
Comment (yangw1234) [reply pyzoo/zoo/pipeline/api/net.py]: remove those imports

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala (hunk 3 above)
Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: added

File: pyzoo/zoo/pipeline/api/net.py (hunk 4 above)
Comment (yangw1234) [reply pyzoo/zoo/pipeline/api/net.py]: removed those imports and use tensorflow.keras instead

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py (hunk 6 above)
Comment (yangw1234) [reply pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: removed this

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py (hunk 7 above)
Comment (yangw1234) [reply pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: removed

**Comment by yangw1234** [comment]
Hi @hkvision can you verify this pr works for nlp architect

File: pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py (hunk 8 above)
Comment (yangw1234) [reply pyzoo/zoo/examples/tensorflow/distributed_training/train_mnist_keras.py]: removed

**Comment by yangw1234** [comment]
jenkins pass http://10.239.47.211:18888/job/ZOO-PR-Validation/1503/

**Comment by hkvision** [comment]
@yangw1234 Sure. But it may take some time for me to migrate nlp-architect. You can merge it first if you wish.

**Comment by yangw1234** [comment]
merge it first and I'll update the document in the next pr

**Comment by hhbyyh** [comment]
With this change, user will meet an error if they have Keras 2.x installed in their system.

```
2018-12-25 17:55:08 WARN  Utils:66 - Your hostname, yuhao-devbox resolves to a loopback address: 127.0.1.1; using 143.183.81.148 instead (on interface enp0s31f6)
2018-12-25 17:55:08 WARN  Utils:66 - Set SPARK_LOCAL_IP if you need to bind to another address
Using TensorFlow backend.
Traceback (most recent call last):
  File "/home/yuhao/workspace/github/hhbyyh/BigDL-ImageProcessing-Examples/xray_nnclassifier/train.py", line 20, in <module>
    from zoo.pipeline.api.net import Net
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/zoo/pipeline/api/net.py", line 35, in <module>
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/bigdl/keras/optimization.py", line 22, in <module>
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/bigdl/keras/converter.py", line 22, in <module>
ImportError: cannot import name 'Layer'
2018-12-25 17:55:08 WARN  Utils:66 - Your hostname, yuhao-devbox resolves to a loopback address: 127.0.1.1; using 143.183.81.148 instead (on interface enp0s31f6)
2018-12-25 17:55:08 WARN  Utils:66 - Set SPARK_LOCAL_IP if you need to bind to another address
Using TensorFlow backend.
Traceback (most recent call last):
  File "/home/yuhao/workspace/github/hhbyyh/BigDL-ImageProcessing-Examples/xray_nnclassifier/train.py", line 20, in <module>
    from zoo.pipeline.api.net import Net
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/zoo/pipeline/api/net.py", line 35, in <module>
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/bigdl/keras/optimization.py", line 22, in <module>
  File "/home/yuhao/workspace/github/hhbyyh/analytics-zoo/dist/lib/analytics-zoo-bigdl_0.7.1-spark_2.1.0-0.4.0-SNAPSHOT-python-api.zip/bigdl/keras/converter.py", line 22, in <module>
ImportError: cannot import name 'Layer'
``

In bigdl/keras/converter.py 
`from keras.models import Sequential, Model, Layer` will cause error since Layer in not there anymore in Keras 2.x.

**Comment by hhbyyh** [comment]
issue opened to track it: https://github.com/intel-analytics/analytics-zoo/issues/1019

**Pull request merged**